src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/IDE/VSSolutionManager.cs

```diff
@@ -41,6 +41,7 @@ public sealed class VSSolutionManager : IVsSolutionManager, IVsSelectionEvents
         private static readonly string VSNuGetClientName = "NuGet VS VSIX";
 
         private readonly INuGetLockService _initLock;
+        private readonly ReentrantSemaphore _semaphoreLock = ReentrantSemaphore.Create(1, NuGetUIThreadHelper.JoinableTaskFactory.Context, ReentrantSemaphore.ReentrancyMode.Freeform);
```

Comment:
In cases where it's used to coordinate only 1 access, and allow switching to the UI thread, you can use Stack reentracy mode. 

I don't think Freeform makes it much worse as it's the most permissive on.